storage/layers.go

```diff
@@ -198,8 +199,37 @@ type DiffOptions struct {
 // stagedLayerOptions are the options passed to .create to populate a staged
 // layer
 type stagedLayerOptions struct {
+	// These are used via the zstd:chunked pull paths
 	DiffOutput  *drivers.DriverWithDifferOutput
 	DiffOptions *drivers.ApplyDiffWithDifferOpts
+
+	// stagedLayerExtraction is used by the normal tar layer extraction.
+	stagedLayerExtraction *maybeStagedLayerExtraction
+}
+
+// maybeStagedLayerExtraction is a helper to encapsulate details around extracting
+// a layer potentially before we even take a look if the driver implements the
+// ApplyDiffStaging interface.
+type maybeStagedLayerExtraction struct {
+	// diff contains the tar archive, can be compressed
+	diff    io.Reader
+	staging drivers.ApplyDiffStaging
+	result  *applyDiffResult
```

Comment:
Later: Please document what is set when / who is responsible for what, if it can’t be expressed by structure of the types.